File: apps/mobile/modules/timer-widget/android/src/main/java/com/actiko/widget/WidgetDbHelper.kt
package com.actiko.widget

import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.util.Log
import java.io.File

data class ActivityRow(
    val id: String,
    val name: String,
    val emoji: String,
    val quantityUnit: String,
    val recordingMode: String,
    val recordingModeConfig: String?,
)

data class KindRow(
    val id: String,
    val name: String,
    val color: String?,
)

class WidgetDbHelper(private val context: Context) {
    companion object {
        const val SUPPORTED_SCHEMA_VERSION = 12

        fun parseCounterSteps(configJson: String?): List<Int> =
            WidgetDbConfigParser.parseCounterSteps(configJson)
    }

    private val dbPath: String
        get() = File(context.filesDir, "SQLite/actiko.db").absolutePath

    internal fun openDatabase(): SQLiteDatabase? {
        val file = File(dbPath)
        if (!file.exists()) {
            Log.e("WidgetDb", "DB not found at: $dbPath")
            return null
        }
        return SQLiteDatabase.openDatabase(
            dbPath,
            null,
            SQLiteDatabase.OPEN_READWRITE,
        ).also {
            it.enableWriteAheadLogging()
            it.execSQL("PRAGMA busy_timeout = 5000")
        }
    }

}

File: apps/mobile/targets/widget/WidgetDbHelper.swift
import Foundation
import SQLite3

private let sqliteTransient = unsafeBitCast(-1, to: sqlite3_destructor_type.self)

enum WidgetDbError: Error {
    case databaseUnavailable
    case prepareFailed(Int32)
    case bindFailed(Int32)
    case stepFailed(Int32)
    case noMatchingRow
}

struct ActivityRow {
    let id: String
    let name: String
    let emoji: String
    let quantityUnit: String
    let recordingMode: String
    let recordingModeConfig: String?
}

struct KindRow {
    let id: String
    let name: String
    let color: String?
}

/// SQLite access to the shared actiko.db via App Group container.
/// iOS port of Android's WidgetDbHelper.kt.
struct WidgetDbHelper {
    static let supportedSchemaVersion = 12
    private let databasePathOverride: String?

    init(databasePath: String? = nil) {
        databasePathOverride = databasePath
    }

    private static var dbPath: String {
        guard let container = FileManager.default.containerURL(
            forSecurityApplicationGroupIdentifier: AppConfig.appGroupId
        ) else { return "" }
        return container.appendingPathComponent("SQLite/actiko.db").path
    }

    func openDatabase() -> OpaquePointer? {
        let path = databasePathOverride ?? Self.dbPath
        guard FileManager.default.fileExists(atPath: path) else { return nil }
        var db: OpaquePointer?
        let openStatus = sqlite3_open_v2(path, &db, SQLITE_OPEN_READWRITE, nil)
        guard openStatus == SQLITE_OK else {
            if let db { sqlite3_close(db) }
            return nil
        }
        sqlite3_exec(db, "PRAGMA journal_mode = WAL;", nil, nil, nil)
        sqlite3_exec(db, "PRAGMA busy_timeout = 5000;", nil, nil, nil)
        return db
    }

    func readActivityRow(_ stmt: OpaquePointer) -> ActivityRow {
        ActivityRow(
            id: columnText(stmt, 0),
            name: columnText(stmt, 1),
            emoji: columnOptionalText(stmt, 2) ?? "",
            quantityUnit: columnOptionalText(stmt, 3) ?? "",
            recordingMode: columnOptionalText(stmt, 4) ?? "timer",
            recordingModeConfig: columnOptionalText(stmt, 5)
        )
    }

    func columnText(_ stmt: OpaquePointer, _ index: Int32) -> String {
        if let cStr = sqlite3_column_text(stmt, index) {
            return String(cString: cStr)
        }
        return ""
    }

    func columnOptionalText(_ stmt: OpaquePointer, _ index: Int32) -> String? {
        if sqlite3_column_type(stmt, index) == SQLITE_NULL { return nil }
        if let cStr = sqlite3_column_text(stmt, index) {
            return String(cString: cStr)
        }
        return nil
    }

    @discardableResult
    func bindText(_ stmt: OpaquePointer, _ index: Int32, _ value: String) -> Int32 {
        sqlite3_bind_text(stmt, index, (value as NSString).utf8String, -1, sqliteTransient)
    }

    @discardableResult
    func bindOptionalText(_ stmt: OpaquePointer, _ index: Int32, _ value: String?) -> Int32 {
        if let value = value {
            return sqlite3_bind_text(
                stmt,
                index,
                (value as NSString).utf8String,
                -1,
                sqliteTransient
            )
        } else {
            return sqlite3_bind_null(stmt, index)
        }
    }
}
